TestKit: Add Shared Swift Package for Test Utilities

shiki opened the pull request:
Ref paaHJt-1gq-p2

This adds a new Swift Package named TestKit. Thanks for everyone's suggestions on the name. 

<img src="https://user-images.githubusercontent.com/198826/91097537-16d94800-e61d-11ea-85db-3e7125ac5a15.png" width="450">

The package is added to the `*Tests` target like this:

![image](https://user-images.githubusercontent.com/198826/91097990-e514b100-e61d-11ea-8c31-427614a93e18.png)

The goal of the package is to house all the general assertions and `XCTExpectation` helpers we have throughout the projects. I have not moved all the assertions yet, just the [`XCTestCase+Wait`](https://github.com/woocommerce/woocommerce-ios/blob/add/tests-shared-spm/TestKit/Sources/TestKit/XCTestCase%2BWait.swift) helpers for now.

## Unusual Importing

You'll notice in the changes that I only had to have one `import TestKit` declaration in every project. For example, for the `WooCommerceTests`, I added it to the `SurveyViewControllerTests` **only**. After that, all other tests classes can access the `waitUntil` methods. 🤷 

https://github.com/woocommerce/woocommerce-ios/blob/9681d6b69f0e8a092854ac627935031100f0575e/WooCommerce/WooCommerceTests/ViewRelated/Survey/SurveyViewControllerTests.swift#L4

## Wait Timeout

The `waitUntil` and `waitForExpectation` helpers now have `1.0` seconds timeouts by default:

https://github.com/woocommerce/woocommerce-ios/blob/9681d6b69f0e8a092854ac627935031100f0575e/TestKit/Sources/TestKit/XCTestCase%2BWait.swift#L17-L20

I think this is a better default instead of `10` seconds. I believe we should strive to have unit tests that execute fast. But let me know if this is a concern for you. The `timeout` can be overridden for every call anyway. 🙂 

## Generated Files

There are some generated files that probably have no bearing but I just left them there for now. 

- [`LinuxMain.swift`](https://github.com/woocommerce/woocommerce-ios/blob/add/tests-shared-spm/TestKit/Tests/LinuxMain.swift)
- [`XCTestsManifests.swift`](https://github.com/woocommerce/woocommerce-ios/blob/add/tests-shared-spm/TestKit/Tests/TestKitTests/XCTestManifests.swift)

## Testing

### Local

Please run all the unit tests locally and confirm that they passed. 

### Circle CI

Please check that these are passing: 

- Unit tests 
- UI tests
- Installable builds

### Isolated Package Build

1. In Terminal, `cd /path/to/project-root`
2. `cd TestKit`
3. `swift package generate-xcodeproj`. This will generate a project file for `TestKit` that can be opened by Xcode.
4. `open TestKit.xcodeproj`. Xcode should have opened `TestsKit.xcodeproj`. 
5. In Xcode, click on Product → Test to run the unit tests. There are no unit tests, this is just to make sure that it builds. 😀 

## Submitter Checklist

- [x] If it's feasible, I have added unit tests. 
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

shiki committed:
Generate "Testing" Swift Package and add to Xcode Project

I also added iOS .v12 as the platforms supported.

shiki committed:
Link Testing package to YosemiteTests target

shiki committed:
Add Testing.waitForExpectation

This was copied from YosemiteTests.

shiki committed:
Delete Yosemite.waitForExpectation

shiki committed:
Import Testing in Yosemite

It looks like I only need to import it to one of the test classes. All other
test classes get the public waitForExpectation method automatically.

shiki committed:
Link Testing to StorageTests target

shiki committed:
Remove Storage's waitForExpectation

shiki committed:
Import Testing library to a Storage test

shiki committed:
Copy waitUntil from WooCommerce package

shiki committed:
Link Testing package to WooCommerceTests target

shiki committed:
Delete XCTestCase+Wait file from WooCommerce

shiki committed:
Import Testing package in SurveyVCTests

shiki committed:
Link Testing package to NetworkingTests target

shiki committed:
Delete XCTestCase+Wait from Networking

shiki committed:
Import Testing in String+HTMLTests

shiki committed:
Delete unused Testing.swift

shiki committed:
Update Testing/README.md

shiki committed:
Testing: Increase timeout to 1.0

Some of our tests are alow in CircleCI. XD

shiki committed:
Remove Testing package from projects

This will be renamed later. Renaming the package directly in Xcode seems to make
things blow up.

shiki committed:
Rename "Testing" folder to "TestKit"

shiki committed:
Add TestKit reference back to workspace

shiki committed:
Update naming in TestKit/Package.swift

shiki committed:
Update naming in TestKit/README.md

shiki committed:
Rename Sources/Testing to Sources/TestKit

shiki committed:
Rename Tests/TestingTests to Tests/TestKitTests

shiki committed:
Rename references to "Testing" to "TestKit"

shiki committed:
Add TestKit to NetworkingTests target

shiki committed:
Add TestKit to StorageTests target

shiki committed:
Add TestKit to WooCommerceTests target

shiki committed:
Add TestKit to YosemiteTests target

shiki committed:
Reorder workspace

shiki committed:
Update imports of "Testing" to use "TestKit"

shiki committed:
SurveyCoordinatorControllerTests: Add timeout for waitUntil

The default timeout is now 1 second but this test takes a bit longer to finish.

shiki committed:
Merge branch develop into add/tests-shared-spm

Ecarrion commented:
Thank for this @shiki, all looks great 🚀  
I do have some concern about:
> The waitUntil and waitForExpectation helpers now have 1.0 seconds timeouts by default:

I think that `wait-until predicate` doest does not work consistently all the times. I had some tests that used a `1.0s` timeout and were causing random failures from time to time. 

I couldn't found out the reason but increasing the timeout fixed the issue 😇 https://github.com/woocommerce/woocommerce-ios/pull/2656

pmusolino reviewed (commented):
Thanks @shiki for your proposal!

I have the same feeling of @Ecarrion. It seems that some tests take, by default, more than 1 second and they depend on the machine loads, then sometimes you can see them failing and other times no.

In my case, running the tests locally, I've seven failing tests, and the failure reason for all the tests is `failed - Timed out waiting for condition to return true.`

<img width="615" alt="Schermata 2020-08-25 alle 17 13 34" src="https://user-images.githubusercontent.com/495617/91192698-87a06300-e6f6-11ea-9e13-0364a4a5b0d5.png">

shiki committed:
Change timeout for wait* helpers to 5.0

We have some tests that fail if the timeout is 1.0. We will have to review this
later if it's a good idea.

shiki commented:
Thank you for checking, @pmusolino. It's so ironic because most of those failing tests are mine. 🤣 

I changed the timeout to 5.0 (0f8cf76). Would you mind testing again if this works on your machine?

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Survey/SurveyCoordinatorControllerTests.swift:
```diff
@@ -43,7 +43,7 @@ final class SurveyCoordinatingControllerTests: XCTestCase {
 
         // Then
-        waitUntil {
+        waitUntil(timeout: 3.0) {
```
just wondering, is this overriding the default timeout 5s because we know this call finishes sooner?

shiki replied:
Thanks for noticing that. I forgot to remove this when I changed the default timeout to 5.0 in 0f8cf76b6eb4e464b1f26811838ddcd34fb1a347. Fixed in 37129f819e8094f964965bf28be158af4d759b51.

jaclync commented on TestKit/Sources/TestKit/XCTestCase+Wait.swift:
```diff
@@ -14,10 +14,12 @@ extension XCTestCase {
-                            timeout: TimeInterval = Constants.expectationTimeout,
+    public func waitForExpectation(description: String? = nil,
+                            count: Int = 1,
+                            timeout: TimeInterval = 5.0,
```
nit: indentation from `public` 🙂

shiki replied:
Thanks! Fixed in 6281cb611ec64e9ed77652769964b901136c4841.

jaclync commented on TestKit/Sources/TestKit/XCTestCase+Wait.swift:
```diff
@@ -34,9 +36,9 @@ extension XCTestCase {
                    line: UInt = #line,
-                   timeout: TimeInterval = Constants.expectationTimeout,
+                   timeout: TimeInterval = 5.0,
                    condition: @escaping (() -> Bool)) {
```
nit: similar to above, these 4 lines could be further indented from `public` 🙂

shiki replied:
Thanks! Fixed in 6281cb611ec64e9ed77652769964b901136c4841.

jaclync reviewed (approved):
Nice work @shiki figuring out integrating a shared framework to the project via SPM!

Side question: do you happen to know if it's easy to add dependencies that are also shared in other frameworks in the project? for example, if we have another shared framework for a test target that depends on `Storage` (internal) and `Alamofire` (external) which are also dependencies for the main app. I'm guessing the external libraries might be tricky because Cocoapods manages the third-party libraries separately?

Code: ✅ 
Testing: ✅ (ran all the tests locally, all passed if the simulator isn't using a non-gregorian calendar!)

jaclync commented:
After checking out another branch from the PR branch, I noticed that Xcode showed an alert:

<img width="494" alt="Screen Shot 2020-08-26 at 4 32 31 PM" src="https://user-images.githubusercontent.com/1945542/91280841-f266a280-e7b9-11ea-92f4-550bc8fc2d5b.png">

And then after tapping "Keep Xcode version" there are diffs in `WooCommerce.xcworkspace/contents.xcworkspacedata`:
```
--- a/WooCommerce.xcworkspace/contents.xcworkspacedata
+++ b/WooCommerce.xcworkspace/contents.xcworkspacedata
@@ -13,6 +13,9 @@
    <FileRef
       location = "group:Networking/Networking.xcodeproj">
    </FileRef>
+   <FileRef
+      location = "group:TestKit">
+   </FileRef>
    <FileRef
       location = "group:Pods/Pods.xcodeproj">
    </FileRef>
```

I was able to repro this twice. This probably wouldn't be an issue once all the branches have `TestKit`?

pmusolino commented:
Thank you @shiki for updating it.
All the tests are green now, but I get the same alert of @jaclync. After switching to this branch, Xcode asks me to use the version on disk or keep the Xcode version. I tried to press `Use version on Disk`, and in my case, I've diff on `WooCommerce.xcodeproj`. 

Diff 1:
```
-		57150E0F24F462C200E81611 /* TestKit in Frameworks */ = {isa = PBXBuildFile; productRef = 57150E0E24F462C200E81611 /* TestKit */; };
+		57150E0F24F462C200E81611 /* BuildFile in Frameworks */ = {isa = PBXBuildFile; productRef = 57150E0E24F462C200E81611 /* SwiftPackageProductDependency */; };
```

Diff 2: 
```
-				57150E0F24F462C200E81611 /* TestKit in Frameworks */,
+				57150E0F24F462C200E81611 /* BuildFile in Frameworks */,
```

Diff 3:
```
-				57150E0E24F462C200E81611 /* TestKit */,
+				57150E0E24F462C200E81611 /* SwiftPackageProductDependency */,
```

Diff 4:
```
-			inputPaths = (
-				"${PODS_ROOT}/Target Support Files/Pods-WooCommerce/Pods-WooCommerce-resources.sh",
-				"${PODS_ROOT}/GoogleSignIn/Resources/GoogleSignIn.bundle",
-				"${PODS_CONFIGURATION_BUILD_DIR}/WordPressAuthenticator/WordPressAuthenticatorResources.bundle",
+			inputFileListPaths = (
+				"${PODS_ROOT}/Target Support Files/Pods-WooCommerce/Pods-WooCommerce-resources-${CONFIGURATION}-input-files.xcfilelist",
```

Diff 5:
```
-			outputPaths = (
-				"${TARGET_BUILD_DIR}/${UNLOCALIZED_RESOURCES_FOLDER_PATH}/GoogleSignIn.bundle",
-				"${TARGET_BUILD_DIR}/${UNLOCALIZED_RESOURCES_FOLDER_PATH}/WordPressAuthenticatorResources.bundle",
+			outputFileListPaths = (
+				"${PODS_ROOT}/Target Support Files/Pods-WooCommerce/Pods-WooCommerce-resources-${CONFIGURATION}-output-files.xcfilelist",
```

Diff 7:
```
-			inputPaths = (
-				"${PODS_ROOT}/Target Support Files/Pods-WooCommerce/Pods-WooCommerce-frameworks.sh",
-				"${BUILT_PRODUCTS_DIR}/1PasswordExtension/OnePasswordExtension.framework",
-				"${BUILT_PRODUCTS_DIR}/Alamofire/Alamofire.framework",
-				"${BUILT_PRODUCTS_DIR}/AppAuth/AppAuth.framework",
-				"${BUILT_PRODUCTS_DIR}/Automattic-Tracks-iOS/AutomatticTracks.framework",
-				"${BUILT_PRODUCTS_DIR}/Charts/Charts.framework",
-				"${BUILT_PRODUCTS_DIR}/CocoaLumberjack/CocoaLumberjack.framework",
-				"${BUILT_PRODUCTS_DIR}/FormatterKit/FormatterKit.framework",
-				"${BUILT_PRODUCTS_DIR}/GTMAppAuth/GTMAppAuth.framework",
-				"${BUILT_PRODUCTS_DIR}/GTMSessionFetcher/GTMSessionFetcher.framework",
-				"${BUILT_PRODUCTS_DIR}/Gridicons/Gridicons.framework",
-				"${BUILT_PRODUCTS_DIR}/KeychainAccess/KeychainAccess.framework",
-				"${BUILT_PRODUCTS_DIR}/Kingfisher/Kingfisher.framework",
-				"${BUILT_PRODUCTS_DIR}/NSObject-SafeExpectations/NSObject_SafeExpectations.framework",
-				"${BUILT_PRODUCTS_DIR}/NSURL+IDN/NSURL_IDN.framework",
-				"${BUILT_PRODUCTS_DIR}/Reachability/Reachability.framework",
-				"${BUILT_PRODUCTS_DIR}/SVProgressHUD/SVProgressHUD.framework",
-				"${BUILT_PRODUCTS_DIR}/Sentry/Sentry.framework",
-				"${BUILT_PRODUCTS_DIR}/Sodium/Sodium.framework",
-				"${BUILT_PRODUCTS_DIR}/UIDeviceIdentifier/UIDeviceIdentifier.framework",
-				"${BUILT_PRODUCTS_DIR}/WPMediaPicker/WPMediaPicker.framework",
-				"${BUILT_PRODUCTS_DIR}/WordPress-Aztec-iOS/Aztec.framework",
-				"${BUILT_PRODUCTS_DIR}/WordPress-Editor-iOS/WordPressEditor.framework",
-				"${BUILT_PRODUCTS_DIR}/WordPressKit/WordPressKit.framework",
-				"${BUILT_PRODUCTS_DIR}/WordPressShared/WordPressShared.framework",
-				"${BUILT_PRODUCTS_DIR}/WordPressUI/WordPressUI.framework",
-				"${BUILT_PRODUCTS_DIR}/Wormholy/Wormholy.framework",
-				"${BUILT_PRODUCTS_DIR}/XLPagerTabStrip/XLPagerTabStrip.framework",
-				"${PODS_ROOT}/ZendeskCommonUISDK/CommonUISDK.framework",
-				"${PODS_ROOT}/ZendeskCoreSDK/ZendeskCoreSDK.framework",
-				"${PODS_ROOT}/ZendeskMessagingAPISDK/MessagingAPI.framework",
-				"${PODS_ROOT}/ZendeskMessagingSDK/MessagingSDK.framework",
-				"${PODS_ROOT}/ZendeskSDKConfigurationsSDK/SDKConfigurations.framework",
-				"${PODS_ROOT}/ZendeskSupportProvidersSDK/SupportProvidersSDK.framework",
-				"${PODS_ROOT}/ZendeskSupportSDK/SupportSDK.framework",
-				"${BUILT_PRODUCTS_DIR}/lottie-ios/Lottie.framework",
-				"${BUILT_PRODUCTS_DIR}/wpxmlrpc/wpxmlrpc.framework",
+			inputFileListPaths = (
+				"${PODS_ROOT}/Target Support Files/Pods-WooCommerce/Pods-WooCommerce-frameworks-${CONFIGURATION}-input-files.xcfilelist",
```

Diff 8:
```
-			outputPaths = (
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/OnePasswordExtension.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Alamofire.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/AppAuth.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/AutomatticTracks.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Charts.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/CocoaLumberjack.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/FormatterKit.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/GTMAppAuth.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/GTMSessionFetcher.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Gridicons.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/KeychainAccess.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Kingfisher.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/NSObject_SafeExpectations.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/NSURL_IDN.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Reachability.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/SVProgressHUD.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Sentry.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Sodium.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/UIDeviceIdentifier.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/WPMediaPicker.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Aztec.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/WordPressEditor.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/WordPressKit.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/WordPressShared.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/WordPressUI.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Wormholy.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/XLPagerTabStrip.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/CommonUISDK.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/ZendeskCoreSDK.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/MessagingAPI.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/MessagingSDK.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/SDKConfigurations.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/SupportProvidersSDK.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/SupportSDK.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Lottie.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/wpxmlrpc.framework",
+			outputFileListPaths = (
+				"${PODS_ROOT}/Target Support Files/Pods-WooCommerce/Pods-WooCommerce-frameworks-${CONFIGURATION}-output-files.xcfilelist",
```

Diff 9:
```
-		57150E0E24F462C200E81611 /* TestKit */ = {
+		57150E0E24F462C200E81611 /* SwiftPackageProductDependency */ = {
```

shiki commented:
@jaclync @pmusolino I did a test by switching to another branch:

With Xcode **closed** while switching to this branch: Git diff looked fine. 

```
On branch add/tests-shared-spm
Your branch is up to date with 'origin/add/tests-shared-spm'.

nothing to commit, working tree clean
```

With Xcode **open** while switching to this branch and choosing "Use Version on Disk". The git status still looked fine. 

```
On branch add/tests-shared-spm
Your branch is up to date with 'origin/add/tests-shared-spm'.

nothing to commit, working tree clean
```

With Xcode **open** while switching to this branch and choosing "Keep Xcode Version". The git diff is what is expected from pressing that button:

```diff
diff --git a/WooCommerce.xcworkspace/contents.xcworkspacedata b/WooCommerce.xcworkspace/contents.xcworkspacedata
index 9bdc0303b..54ee0fedf 100644
--- a/WooCommerce.xcworkspace/contents.xcworkspacedata
+++ b/WooCommerce.xcworkspace/contents.xcworkspacedata
@@ -13,9 +13,6 @@
    <FileRef
       location = "group:Networking/Networking.xcodeproj">
    </FileRef>
-   <FileRef
-      location = "group:TestKit">
-   </FileRef>
    <FileRef
       location = "group:Pods/Pods.xcodeproj">
    </FileRef>
```

--- 

I'm using Xcode 11.6 by the way. I'm not sure if that matters. 

Anyway, I think all issues you mentioned are expected and shouldn't be a problem. I've run into similar issues with Xcode before but switching branches while Xcode is not open should fix things.

shiki committed:
SurveyCoordinatorControllerTests: Remove waitUntil timeout

The default is currently 5.0 so this isn't really necessary.

shiki committed:
XCTestCase+Wait: Fix indentations

shiki commented:
Thank you, everyone, for testing this. ❤️ I will merge this now. Please let me know if something breaks. 🙈

mokagio commented:
> Anyway, I think all issues you mentioned are expected and shouldn't be a problem. I've run into similar issues with Xcode before but switching branches while Xcode is not open should fix things.

Xcode gets pretty confused when the branch is changed under its feet. It crashed on me quite a few times doing stuff like this in the past.
